atomics/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,427 @@
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T old;
+#pragma acc atomic capture
+  {
+    old = *ptr;
+    *ptr ^= val;
+  }
+  return old;
+}
+//</editor-fold>
+
+//<editor-fold
+// desc="device_atomic_{add,sub,mul,div,lshift,rshift,mod,max,min,and,or,xor}_fetch">
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_add_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr += val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_inc_fetch(
+    T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr += T(1);
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_sub_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr -= val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_dec_fetch(
+    T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr -= T(1);
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_mul_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr *= val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_div_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr /= val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_integral_type_v<T>, T> device_atomic_lshift_fetch(
+    T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr = *ptr << val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_integral_type_v<T>, T> device_atomic_rshift_fetch(
+    T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  {
+    *ptr = *ptr >> val;
+    tmp = *ptr;
+  }
+  return tmp;
+}
+
+#ifdef __NVCOMPILER
+#pragma acc routine seq
+template <class T>
+std::enable_if_t<is_openacc_arithmetic_type_v<T>, T> device_atomic_max_fetch(
+    T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  tmp = atomicMax(ptr, val);
+  tmp = std::max(tmp, val);
```

Comment:
?? signing twice why? Silencing warning for unused return argument?

Reply:
The `atomic_max_fetch()` should return (fetch) the result of the max operation, but `atomicMax()` returns old value pointed by `ptr`. Therefore, we must again calculate `max(old-value, val)` to return it.

Reply:
An alternative option will be to return `*ptr` instead of re-calculating it, but I'm not sure which will be better.

Reply:
> An alternative option will be to return `*ptr` instead of re-calculating it, but I'm not sure which will be better.

No that would be a race condition.

Wha you do here is fine.

Reply:
You are correct; I forgot why I did this way.